File: 1_TwoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
//解法1
const twoSum = function (nums, target) {
    //結果的陣列
    let result = [];
    //跑巢狀迴圈加總  加總=target  回傳結果
    for (let i = 0; i < nums.length; i++) {
        for (let j = i + 1; j < nums.length; j++) {
            if ((nums[i] + nums[j]) === target) {
                result.push(i);
                result.push(j);
                return result;
            }
        }
    }
    return result;
};

//解法2
const twoSumV2 = (nums, target) => {
    //使用obj  可將巢狀迴圈縮減
    let map = {};

    for(let i = 0; i < nums.length; i++){
        //將target - 對應元素
        let minus = target - nums[i];
        
        //若是相減完的結果在map內  return結果
        if(minus in map){
            return[map[minus], i]
        }
        //若未比對到  key為值 value為他在nums內的索引
        map[nums[i]] = i;
    }
    return null;
}

File: 1.TwoSum/1_TwoSum.py
# 解法1
def twoSum(nums: list[int], target: int) -> list[int]:
    # 要回傳的結果陣列
    result = []

    # 巢狀迴圈加總與target比對
    for index in range(0, len(nums)):

        # 內部迴圈要跳過上層已遍歷過的index
        for index1 in range(index+1, len(nums)):
            if (nums[index] + nums[index1]) == target:
                result.append(index)
                result.append(index1)

    return result

# 解法2


def twoSum1(nums: list[int], target: int) -> list[int]:
    # 使用dic  key儲存nums[i]的值 value儲存index
    map = {}
    for index in range(len(nums)):
        # target - 當前元素
        minus = target - nums[index]

        # 相減完結果在map內即回傳
        if minus in map:
            return [map[minus], index]
        else:
            map[nums[index]] = index
